app/queries/get_application_choices_for_providers.rb

```diff
@@ -12,40 +12,32 @@ class GetApplicationChoicesForProviders
   ].freeze
 
   def self.call(providers:, vendor_api: false, includes: DEFAULT_INCLUDES, recruitment_cycle_year: RecruitmentCycle.years_visible_to_providers)
-    providers = Array.wrap(providers).select(&:present?)
-
-    raise MissingProvider if providers.none?
+    raise MissingProvider if providers.blank? # super important!
 
+    provider_ids = providers.map(&:id)
     statuses = vendor_api ? ApplicationStateChange.states_visible_to_provider_without_deferred : ApplicationStateChange.states_visible_to_provider
 
-    with_course_joins = ApplicationChoice
-      .joins('INNER JOIN course_options AS current_course_option ON current_course_option_id = current_course_option.id')
-      .joins('INNER JOIN course_options AS original_option ON course_option_id = original_option.id')
-      .joins('INNER JOIN courses AS current_course ON current_course_option.course_id = current_course.id')
-      .joins('INNER JOIN courses AS original_course ON original_option.course_id = original_course.id')
-
-    applications =
-      with_course_joins.where(
-        'original_course.provider_id' => providers,
-        'original_course.recruitment_cycle_year' => recruitment_cycle_year,
```

Comment:
we seem to be looking for the original course's `recruitment_cycle_year`,  but now we're only capturing the current recruitment cycle year from the current option, will this pose any issues? (you prob mentioned something around this and I forgot)

Reply:
We have recently changed the behaviour of the filters within Manage to always reference the current year, as defined by the current course associated with the application choice. Therefore, the original provider cannot actually look up an application using the year of the original course, but may do using the year of the current one. This, of course, is only relevant if the years differ, which is only possible in a defer/re-instate scenario.

Filtering aside, visibility of applications is restricted to the current and previous recruitment cycle years, which means, in both versions of the code, an application from the previous year is always visible to the original provider. But with the new version of the service, such applications stay visible to their original providers for longer: changes to courses will either preserve the current year or increment it and visibility will be based on the current year.

To give an example: candidate applied in 2020, their placement was deferred and reinstated in the 2021 (current) recruitment cycle, which means the original year is 2020 and the current year is 2021. We'll soon be in year 2022.

When this happens, with our current code this application will disappear for the original provider, as the original year is 2020 and we'll be in 2022. But with the proposed changes, this application will still be visible, and the new code only checks the current year of the application, which is 2021.

So, I think the proposed change is better at avoiding data loss in API responses etc, not worse as I initially thought.

Let me know what you think.

Reply:
I agree with the above, however I think we have one extra point to account for: pending conditions from previous cycle. That might be the only one we might be losing.

Reply:
This scenario is specific to the sorting service. We discussed it on Slack and established: a. there is no data loss at any time in terms of application visibility (GetApplicationChoicesForProviders) and b. the old sorting code already uses the recruitment cycle year of the current course, so we are not introducing any changes there.